starwars_api/models.py

```diff
@@ -55,27 +63,64 @@ def __repr__(self):
 class BaseQuerySet(object):
 
     def __init__(self):
-        pass
+        self.next_page = None
+        self.previous_page = None
+        self.results = []
+        self.current_record = 0
+        self.current_page = 0
 
     def __iter__(self):
-        pass
+        return self
 
     def __next__(self):
         """
         Must handle requests to next pages in SWAPI when objects in the current
         page were all consumed.
         """
-        pass
-
+        while True:
+            if not self.results:
+                self.current_page += 1
+                self.fetch_data(self.current_page)
+                self.current_record = 0
+            try:
+                return_data = self.results[self.current_record]
+            except IndexError:
+                self.current_page += 1
+                try:
+                    self.fetch_data(self.current_page)
+                except SWAPIClientError:
+                    raise StopIteration
+                self.current_record = 0
+                return_data = self.results[self.current_record]
+            self.current_record += 1
```

Comment:
for readability - could put a blank line after this to show that the giant nested code block has ended. and that the next block is an if/else method that's outside the previous if/else method.